core/transaction-pool/graph/src/ready.rs

```diff
@@ -397,14 +413,15 @@ impl<'a, Hash: 'a + hash::Hash + Member, Ex: 'a> BestIterator<'a, Hash, Ex> {
 	}
 }
 
-impl<'a, Hash: 'a + hash::Hash + Member, Ex: 'a> Iterator for BestIterator<'a, Hash, Ex> {
+impl<Hash: hash::Hash + Member, Ex> Iterator for BestIterator<Hash, Ex> {
 	type Item = Arc<Transaction<Hash, Ex>>;
 
 	fn next(&mut self) -> Option<Self::Item> {
 		let best = self.best.iter().next_back()?.clone();
 		let best = self.best.take(&best)?;
 
-		let ready = match self.all.get(&best.transaction.hash) {
+		let next = self.all.read().get(&best.transaction.hash).cloned();
+		let ready = match next {
 			Some(ready) => ready,
 			// The transaction is not in all, maybe it was removed in the meantime?
 			None => return self.next(),
```

Comment:
Out of scope: probably consider replace recursion with `loop`? Stack limit could hit, for example, when pool is full of ready (best) invalid transactions + we're composing a block + there's an active iterator which have the same `best` => `remove_invalid(best)` will be called && all tx from the `best` will be invalidated => this would lead to recursion of `best.len()` depth here.